### DISCUSSION TITLE
Does this library STUN bypass cone/port restricted NAT?

### DISCUSSION THREAD

**Discussion by makew0rld** [pion/webrtc#1789] [Q&A] [answered]
From the [STUN Wikipedia page](https://en.wikipedia.org/wiki/STUN):

> STUN does work with three types of NAT: full cone NAT, restricted cone NAT, and port restricted cone NAT. In the cases of restricted cone or port restricted cone NATs, the client must send out a packet to the endpoint before the NAT will allow packets from the endpoint through to the client.

Does this library handle the case at the end, where clients must send out a packet to the endpoint? I would assume it does, but I'm not sure if this is part of any spec or not. If the library doesn't currently do this I think it should.

Thanks

**Comment by Sean-Der** [comment, marked as answer]
Yes it does! Those terms aren't what people really use anymore. I have attempted to update this page and the WebRTC one and always get reverted. [RFC 4787](https://tools.ietf.org/html/rfc4787#section-4.1) is a great read. It breaks it down into Mapping/Filtering and then the behaviors are address independent/dependent.

At least one peer must be behind a address independent NAT. This allows them to distribute their NAT mapping so others can send traffic in. If both Peers are behind Address Dependent NATs they can't use STUN and have to use a TURN Server.

* 2 Address Independent. Both Peers can share their NAT Mappings and it will work.
* 1 Address Independent. The Independent side can share their mapping, the act of the dependent side contacting the independent peer will create a new mapping.
* 2 Address Dependent. No useful NAT Mapping can be created. The STUN server can respond to your request, but no one else can contact you.

**Reply by makew0rld** [reply]
Great to hear! Thanks for the explanation.

**Reply by ar13101085** [reply]
Hello @Sean-Der ,  I have an observation regarding this WebRTC in my project. I tried connecting two peers behind port-restricted cone NATs, but they couldn't establish a connection with each other. However, I noticed that https://sharedrop.io/ can successfully establish P2P connections in the same network conditions, which I verified using Wireshark. Could you help me understand why this difference exists?"
